homeassistant/components/websocket_api/http.py

```diff
@@ -126,44 +131,34 @@ async def _writer(
         message_queue = self._message_queue
         logger = self._logger
         wsock = self._wsock
-        loop = self._hass.loop
+        loop = self._loop
         debug = logger.debug
         is_enabled_for = logger.isEnabledFor
         logging_debug = logging.DEBUG
         # Exceptions if Socket disconnected or cancelled by connection handler
         try:
             while not wsock.closed:
-                if (messages_remaining := len(message_queue)) == 0:
+                if not message_queue:
                     self._ready_future = loop.create_future()
                     await self._ready_future
-                    messages_remaining = len(message_queue)
 
-                # A None message is used to signal the end of the connection
-                if (message := message_queue.popleft()) is None:
+                if self._closing:
                     return
 
                 debug_enabled = is_enabled_for(logging_debug)
-                messages_remaining -= 1
-
                 if (
-                    not messages_remaining
-                    or not (connection := self._connection)
+                    not (connection := self._connection)
                     or not connection.can_coalesce
+                    or len(message_queue) == 1
                 ):
+                    message = message_queue.popleft()
                     if debug_enabled:
                         debug("%s: Sending %s", self.description, message)
                     await send_bytes_text(message)
                     continue
 
-                messages: list[bytes] = [message]
```

Comment:
Since it's all asyncio and there is no thread safety concern with a separate flag, we can avoid the loop and None check here and simply send the queue